frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -143,35 +163,113 @@ public static ResourceDescriptor fromResource(Resource res) throws Exception {
     } catch (IllegalAccessException e) {
       LOGGER.logError(e, "Ignore: Fail to get GPU information, illegal access function");
     }
+
+    try {
+
+      Class<?> clazz = res.getClass();
+      Class hadoopValueRangesClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRanges");
+      Class hadoopValueRangeClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRange");
+      Method getPorts = clazz.getMethod("getPorts");
+
+      Object hadoopValueRanges = getPorts.invoke(res);
+      if (hadoopValueRanges != null) {
+        Method getBegin = hadoopValueRangeClass.getMethod("getBegin");
+        Method getEnd = hadoopValueRangeClass.getMethod("getEnd");
+
+        Method getSortedRangesList = hadoopValueRangesClass.getMethod("getSortedRangesList");
+        List<Object> hadoopValueRangeList = (List<Object>) getSortedRangesList.invoke(hadoopValueRanges);
+
+        List<Range> rangeList = new ArrayList();
+        for (Object hadoopRange : hadoopValueRangeList) {
+          Range range = new Range();
+          range.setBegin((int) getBegin.invoke(hadoopRange));
+          range.setEnd((int) getEnd.invoke(hadoopRange));
+          LOGGER.logDebug("Get Range: " + range);
+          rangeList.add(range);
```

Comment:
remove

Reply:
?

---
In reply to: [167154655](https://github.com/Microsoft/pai/pull/318#discussion_r167154655) [](ancestors = 167154655)